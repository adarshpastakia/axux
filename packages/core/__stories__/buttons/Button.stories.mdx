import {
  Anchor,
  ArgsTable,
  Canvas,
  Description,
  Heading,
  Meta,
  Story,
  Title
} from "@storybook/addon-docs";
import { AxButton, COLOR } from "../../src";
import {
  ButtonStory,
  OnlyIconStory,
  WithBadgeStory,
  WithIconStory,
  WithTooltipStory
} from "./ButtonStory";
import { mdiBell } from "@mdi/js";

<Meta
  title="@core/Buttons/Button"
  component={AxButton}
  parameters={{ jest: ["Button.test.tsx"] }}
/>

<Anchor storyId="core-buttons-button--button">
  <Title>Button</Title>
</Anchor>

<Description of={AxButton} />

<Canvas>
  <Story name="Button">{ButtonStory}</Story>
</Canvas>

<Anchor storyId="core-buttons-button--with-icon">
  <Heading>With Icon</Heading>
</Anchor>

<Canvas>
  <Story name="With Icon">{WithIconStory}</Story>
</Canvas>

<Anchor storyId="core-buttons-button--only-icon">
  <Heading>Icon Only</Heading>
</Anchor>

<Canvas>
  <Story name="Only Icon">{OnlyIconStory}</Story>
</Canvas>

<Anchor storyId="core-buttons-button--with-tooltip">
  <Heading>With Tooltip</Heading>
</Anchor>

<Canvas>
  <Story name="With Tooltip">{WithTooltipStory}</Story>
</Canvas>

<Anchor storyId="core-buttons-button--with-badge">
  <Heading>With Badge</Heading>
</Anchor>

<Canvas>
  <Story name="With Badge">{WithBadgeStory}</Story>
</Canvas>

<Anchor storyId="core-buttons-button--styles">
  <Heading>Styles</Heading>
</Anchor>

<Canvas>
  <Story name="Styles">
    <div>
      <AxButton type="default">Default</AxButton>
      <AxButton type="outline">Outline</AxButton>
      <AxButton type="solid">Solid</AxButton>
      <AxButton type="link">Link</AxButton>
    </div>
    <div>
      <AxButton data-popover="true" type="default">
        Default
      </AxButton>
      <AxButton data-popover="true" type="outline">
        Outline
      </AxButton>
      <AxButton data-popover="true" type="solid">
        Solid
      </AxButton>
      <AxButton data-popover="true" type="link">
        Link
      </AxButton>
    </div>
    <div>
      <AxButton split data-popover="true" type="default">
        Default
      </AxButton>
      <AxButton split data-popover="true" type="outline">
        Outline
      </AxButton>
      <AxButton split data-popover="true" type="solid">
        Solid
      </AxButton>
      <AxButton split data-popover="true" type="link">
        Link
      </AxButton>
    </div>
  </Story>
</Canvas>

<Anchor storyId="core-buttons-button--icon-styles">
  <Heading>Icon Styles</Heading>
</Anchor>

<Canvas>
  <Story name="Icon Styles">
    <AxButton icon={mdiBell}>Button</AxButton>
    <AxButton icon={mdiBell} iconAlign="end">
      Button
    </AxButton>
    <AxButton icon={mdiBell} iconHilight>
      Button
    </AxButton>
    <AxButton icon={mdiBell} iconHilight iconAlign="end">
      Button
    </AxButton>
    <AxButton type="solid" icon={mdiBell} iconHilight>
      Button
    </AxButton>
    <AxButton type="solid" icon={mdiBell} iconHilight iconAlign="end">
      Button
    </AxButton>
  </Story>
</Canvas>

<Anchor storyId="core-buttons-button--sizes">
  <Heading>Sizes</Heading>
</Anchor>

<Canvas>
  <Story name="Sizes">
    <AxButton size="sm">Small</AxButton>
    <AxButton size="default">Default</AxButton>
    <AxButton size="md">Medium</AxButton>
    <AxButton size="lg">Large</AxButton>
  </Story>
</Canvas>

<Anchor storyId="core-buttons-button--themes">
  <Heading>Themes</Heading>
</Anchor>

<Canvas>
  <Story name="Themes">
    <div>
      {Object.values(COLOR).map((color) => (
        <AxButton key={color} color={color}>
          {color}
        </AxButton>
      ))}
    </div>
    <div>
      {Object.values(COLOR).map((color) => (
        <AxButton key={color} type="solid" color={color}>
          {color}
        </AxButton>
      ))}
    </div>
  </Story>
</Canvas>

<Anchor storyId="core-buttons-button--variants">
  <Heading>Variants</Heading>
</Anchor>

<Canvas>
  <Story name="Variants">
    <AxButton.Positive>Positive Button</AxButton.Positive>
    <AxButton.Negative>Negative Button</AxButton.Negative>
    <AxButton.Neutral>Neutral Button</AxButton.Neutral>
    <AxButton.Action message="Done!">Action Button</AxButton.Action>
    <AxButton.Confirm message="Are you sure?">Confirm Button</AxButton.Confirm>
  </Story>
</Canvas>

<Anchor storyId="core-buttons-button--states">
  <Heading>States</Heading>
</Anchor>

<Canvas>
  <Story name="States">
    <AxButton>Normal</AxButton>
    <AxButton isLoading>Loading</AxButton>
    <AxButton isDisabled>Disabled</AxButton>
  </Story>
</Canvas>

<Heading>Props</Heading>

<ArgsTable of={AxButton} sort="requiredFirst" />
