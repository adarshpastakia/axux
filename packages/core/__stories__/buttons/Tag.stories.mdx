import {
  Anchor,
  ArgsTable,
  Canvas,
  Description,
  Heading,
  Meta,
  Story,
  Title
} from "@storybook/addon-docs";
import { AxFlexBox, AxHotKeyLabel, AxTag, COLOR, PALETTE } from "../../src";
import {
  Clickable,
  Hotkeys, Removable,
  TagSizes,
  TagStory,
  TagThemes,
  WithBadgeStory,
  WithIconStory,
  WithTooltipStory
} from "./TagStory";
import { mdiBell } from "@mdi/js";

<Meta title="@core/Buttons/Tag" component={AxTag} parameters={{ jest: ["Tag.test.tsx"] }} />

<Anchor storyId="core-buttons-tag--tag">
  <Title>Tag</Title>
</Anchor>

<Description of={AxTag} />

<Canvas>
  <Story name="Tag">{TagStory}</Story>
  <Story name="With Icon">{WithIconStory}</Story>
  <Story name="With Tooltip">{WithTooltipStory}</Story>
  <Story name="With Badge">{WithBadgeStory}</Story>
  <Story name="Clickable">{Clickable}</Story>
  <Story name="Removable">{Removable}</Story>
</Canvas>

<Anchor storyId="core-buttons-tag--hotkey">
  <Heading>Hotkey Label</Heading>
</Anchor>

<Canvas>
  <Story name="Hotkey">{Hotkeys.bind({})}</Story>
</Canvas>

<Anchor storyId="core-buttons-tag--sizes">
  <Heading>Sizes</Heading>
</Anchor>

<Canvas>
  <Story name="Sizes">{TagSizes.bind({})}</Story>
</Canvas>

<Anchor storyId="core-buttons-tag--themes">
  <Heading>Themes</Heading>
</Anchor>

<Canvas>
  <Story name="Themes">{TagThemes.bind({})}</Story>
</Canvas>

<Heading>Props</Heading>

<ArgsTable components={{ AxTag, AxHotKeyLabel }} sort="requiredFirst" />
