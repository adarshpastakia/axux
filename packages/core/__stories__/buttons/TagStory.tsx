// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { Story } from "@storybook/react";
import { PropsWithChildren } from "react";
import { AxTag, TagProps } from "../../src/buttons/Tag";

const Template: Story<PropsWithChildren<TagProps>> = (props) => <AxTag {...props} />;

export const TagStory = Template.bind({});
TagStory.args = {
  children: "Tag"
};

export const WithIconStory = Template.bind({});
WithIconStory.args = {
  icon: "mdi mdi-bell",
  children: "Tag"
};

export const WithTooltipStory = Template.bind({});
WithTooltipStory.args = {
  children: "Tag",
  tooltip: {
    placement: "bottom",
    content: "Test tooltip"
  }
};

export const WithBadgeStory = Template.bind({});
WithBadgeStory.args = {
  children: "Tag",
  badge: 9
};

export default { title: "Example/Tag", component: AxTag };
