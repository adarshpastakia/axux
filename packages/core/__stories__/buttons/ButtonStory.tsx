// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { action } from "@storybook/addon-actions";
import { Story } from "@storybook/react";
import { PropsWithChildren } from "react";
import { StateLabel } from "../../../../storybook/components";
import { AxButton, AxFlexBox, AxMenu, COLOR } from "../../src";
import { ButtonProps } from "../../src/buttons/Button";

export const ButtonTemplate: Story<PropsWithChildren<ButtonProps>> = (props) => (
  <AxButton {...props} />
);

export const ButtonStyles: Story<Omit<ButtonProps, "type">> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      {["default", "outline", "solid", "link"].map((type: AnyObject) => (
        <AxFlexBox.Col flex="auto" key={type}>
          <StateLabel>{type}</StateLabel>
          <AxButton {...props} type={type}>
            Click Me!
          </AxButton>
        </AxFlexBox.Col>
      ))}
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonSizes: Story<Omit<ButtonProps, "size">> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      {[
        ["sm", "Small"],
        ["default", "Default"],
        ["md", "Medium"],
        ["lg", "Large"]
      ].map(([size, label]: AnyObject) => (
        <AxFlexBox.Col flex="auto" key={size}>
          <StateLabel>{label}</StateLabel>
          <AxButton {...props} size={size}>
            Click Me!
          </AxButton>
        </AxFlexBox.Col>
      ))}
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonThemes: Story<Omit<ButtonProps, "color">> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      {Object.values(COLOR).map((color) => (
        <AxFlexBox.Col flex="auto" key={color}>
          <StateLabel>{color}</StateLabel>
          <AxButton {...props} color={color}>
            {color}
          </AxButton>
        </AxFlexBox.Col>
      ))}
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const DropdownStory: Story<Omit<ButtonProps, "split" | "children">> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      <AxFlexBox.Col flex="auto">
        <AxButton.Dropdown label="Click Me!" {...props} split={false}>
          <AxMenu>
            <AxMenu.Item label="Dropdown menu" />
          </AxMenu>
        </AxButton.Dropdown>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <AxButton.Dropdown label="Click Me!" {...props} split="Button dropdown">
          <AxMenu>
            <AxMenu.Item label="Dropdown menu" />
          </AxMenu>
        </AxButton.Dropdown>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const IconStyles: Story<Omit<ButtonProps, "icon" | "iconAlign" | "iconHilight">> = (
  props
) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Align start</StateLabel>
        <AxButton {...props} icon="mdi mdi-bell">
          Button
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Align end</StateLabel>
        <AxButton {...props} icon="mdi mdi-bell" iconAlign="end">
          Button
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Hilight end</StateLabel>
        <AxButton {...props} icon="mdi mdi-bell" iconHilight>
          Button
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Hilight start</StateLabel>
        <AxButton {...props} icon="mdi mdi-bell" iconHilight iconAlign="end">
          Button
        </AxButton>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonVariants: Story<ButtonProps> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Positive</StateLabel>
        <AxButton.Positive {...props}>Click Me!</AxButton.Positive>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Negative</StateLabel>
        <AxButton.Negative {...props}>Click Me!</AxButton.Negative>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Neutral</StateLabel>
        <AxButton.Neutral {...props}>Click Me!</AxButton.Neutral>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Action</StateLabel>
        <AxButton.Action {...props} message="Action completed!">
          Click Me!
        </AxButton.Action>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Confirm</StateLabel>
        <AxButton.Confirm {...props} message="Confirm action?" onClick={action("Confirm:onClick")}>
          Click Me!
        </AxButton.Confirm>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonStates: Story<Omit<ButtonProps, "isLoading" | "isDisabled">> = (props) => (
  <AxFlexBox>
    <AxFlexBox.Row>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Normal</StateLabel>
        <AxButton {...props}>Click Me!</AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Loading</StateLabel>
        <AxButton {...props} isLoading>
          Click Me!
        </AxButton>
      </AxFlexBox.Col>
      <AxFlexBox.Col flex="auto">
        <StateLabel>Disabled</StateLabel>
        <AxButton {...props} isDisabled>
          Click Me!
        </AxButton>
      </AxFlexBox.Col>
    </AxFlexBox.Row>
  </AxFlexBox>
);

export const ButtonStory = ButtonTemplate.bind({});
ButtonStory.args = {
  children: "Click Me!"
};

export const WithIconStory = ButtonTemplate.bind({});
WithIconStory.args = {
  icon: "mdi mdi-bell",
  children: "Click Me!"
};

export const OnlyIconStory = ButtonTemplate.bind({});
OnlyIconStory.args = {
  "aria-label": "Icon button",
  icon: "mdi mdi-bell"
};

export const WithTooltipStory = ButtonTemplate.bind({});
WithTooltipStory.args = {
  children: "Tooltip!",
  tooltip: {
    placement: "bottom",
    content: "This is a clickable button"
  }
};

export const WithBadgeStory = ButtonTemplate.bind({});
WithBadgeStory.args = {
  children: "Click Me!",
  badge: {
    color: "danger",
    value: 9
  }
};

export default { title: "Example/Button", component: AxButton };
