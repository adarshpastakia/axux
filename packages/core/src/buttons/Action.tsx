/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

import { isString } from "@axux/utilities/src";
import { useCallback, useEffect, useMemo, useState, type FC } from "react";
import { AxAnimation } from "../animations";
import { AxTooltip } from "../overlays/Tooltip";
import { AxButton, type ButtonProps } from "./Button";

export interface ActionProps extends ButtonProps {
  /**
   * action message
   */
  message: string;
  /**
   * action type
   */
  actionType?: "danger" | "success";
}

/**
 * Action button to show tooltip message on success
 *
 * @requires This button expects a `onClick` return value of boolean type,
 * if true will display given message and display animated check/cross
 *
 * @prop message
 * @prop actionType - danger | success
 * @prop see Button
 */
export const ActionButton: FC<ActionProps> = ({
  onClick,
  message,
  color,
  variant: style,
  tooltip,
  actionType = "success",
  children,
  ...props
}) => {
  const [actionDone, setActionDone] = useState(false);

  /** ***************** handle click *******************/
  const doClick = useCallback(() => {
    const ret = onClick?.();
    void Promise.resolve(ret).then((b) => {
      if (b !== false) {
        setActionDone(true);
      }
    });
  }, [onClick]);

  /** ***************** reset done state on timeout *******************/
  useEffect(() => {
    if (actionDone) {
      const timer = setTimeout(() => setActionDone(false), 2000);
      return () => clearTimeout(timer);
    }
  }, [actionDone]);

  /** ***************** normalize tooltip content *******************/
  const tooltipContent = useMemo(() => {
    if (actionDone || !tooltip) return { content: message };
    return isString(tooltip) ? { content: tooltip } : tooltip;
  }, [message, tooltip, actionDone]);

  /** ***************** component *******************/
  return (
    <AxTooltip
      {...tooltipContent}
      isDisabled={!tooltip}
      isOpen={actionDone || undefined}
      placement="top"
    >
      <AxButton
        {...props}
        onClick={doClick}
        data-active={actionDone}
        data-action-done={actionDone}
        color={actionDone ? actionType : color}
        variant={actionDone ? "solid" : style}
        data-extra={
          actionDone && (
            <div className="animated-svg">
              {actionType === "success" && (
                <AxAnimation.Check className="text-success-base" />
              )}
              {actionType === "danger" && (
                <AxAnimation.Cross className="text-danger-base" />
              )}
            </div>
          )
        }
      >
        {children}
      </AxButton>
    </AxTooltip>
  );
};
