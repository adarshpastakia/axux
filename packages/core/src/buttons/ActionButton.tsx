// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2022
// @license   : MIT

import { FC, MouseEvent, useCallback, useEffect, useState } from "react";
import { AxTooltip } from "../overlays/Tooltip";
import { AppIcons } from "../types/appIcons";
import { AxButton, ButtonProps } from "./Button";

export interface ActionProps extends ButtonProps {
  message: string;
}

export const AxActionButton: FC<ActionProps> = ({ onClick, message, color, icon, ...props }) => {
  const [actionDone, setActionDone] = useState(false);
  const doClick = useCallback(
    (e: MouseEvent) => {
      onClick && onClick(e);
      setActionDone(true);
    },
    [onClick]
  );
  useEffect(() => {
    if (actionDone) {
      const timer = setTimeout(() => setActionDone(false), 2000);
      return () => clearTimeout(timer);
    }
  }, [actionDone]);
  return (
    <AxTooltip content={message} isDisabled={!actionDone} isOpen={actionDone} usePortal>
      <AxButton
        onClick={doClick}
        data-active={actionDone}
        color={actionDone ? "success" : color}
        icon={actionDone ? AppIcons.iconCheckAll : icon}
        {...props}
      />
    </AxTooltip>
  );
};
