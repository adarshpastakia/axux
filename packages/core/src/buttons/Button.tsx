/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */

import { handleClick } from "@axux/utilities/dist/handlers";
import { FC, forwardRef, ForwardRefExoticComponent, useMemo } from "react";
import { AxAnimation } from "../animations";
import { Link, LinkProps } from "../components/Link";
import { useBadge } from "../hooks/useBadge";
import { getTooltipProps } from "../hooks/useTooltip";
import { AxHotKey } from "../hotkeys/HotKey";
import { AxIcon } from "../icons/Icon";
import {
  BadgeType,
  ChildrenProp,
  Color,
  ElementProps,
  EmptyCallback,
  IconProp,
  RefProp,
  Size,
  TooltipType,
} from "../types";
import { AppIcons } from "../types/appIcons";
import { Ellipsis } from "../typography/Ellipsis";
import { ActionButton } from "./Action";
import { ConfirmButton } from "./Confirm";
import { DropdownButton } from "./Dropdown";

export interface ButtonProps extends ElementProps, IconProp, RefProp {
  children?: string;
  /**
   * size
   */
  size?: Size;
  /**
   * color
   */
  color?: Color | "invert";
  /**
   * button style
   */
  variant?: "outline" | "solid" | "link";
  /**
   * icon alignment
   */
  iconAlign?: "start" | "end" | "top";
  /**
   * button type
   */
  type?: "button" | "submit" | "reset";
  /**
   * full width
   */
  fullWidth?: boolean;
  /**
   * active state
   */
  isActive?: boolean;
  /**
   * rounded button
   */
  isRound?: boolean;
  /**
   * loading state
   */
  isLoading?: boolean;
  /**
   * disabled state
   */
  isDisabled?: boolean;
  /**
   * invert text color
   */
  invertColor?: boolean;
  /**
   * show dropdown caret
   */
  showCaret?: boolean;
  /**
   * use loading spinner
   */
  useSpinner?: boolean;
  /**
   * disable tab focus
   */
  noTabFocus?: boolean;
  /**
   * tooltip
   */
  tooltip?: TooltipType;
  /**
   * badge
   */
  badge?: BadgeType;
  /**
   * hot key
   */
  hotKey?: string;
  /**
   * dont propagate click event
   */
  stopPropagation?: boolean;
  /**
   * navigation props
   */
  nav?: LinkProps;

  onClick?: EmptyCallback;
}

/**
 * Action button
 */
export const AxButton: ForwardRefExoticComponent<ButtonProps> & {
  Action: typeof ActionButton;
  Confirm: typeof ConfirmButton;
  Dropdown: typeof DropdownButton;
  Group: FC<ChildrenProp & ElementProps & { isVertical?: boolean }>;
} = forwardRef<HTMLElement, ButtonProps>((props, ref) => {
  const {
    icon,
    variant: style,
    size,
    children,
    type = "button",
    color,
    rtlFlip,
    noTabFocus,
    iconAlign,
    className,
    isActive = false,
    isRound = false,
    isLoading = false,
    isDisabled = false,
    showCaret,
    tooltip,
    badge,
    fullWidth,
    invertColor,
    hotKey,
    onClick,
    useSpinner,
    stopPropagation = false,
    // @ts-expect-error
    "data-extra": extra,
    // @ts-expect-error
    "data-popover-open": popoverOpen,
    ...rest
  } = props;

  const Badge = useBadge(badge);
  const isSquare = useMemo(() => !!icon && !children, [icon, children]);

  const tooltipProps = useMemo(() => getTooltipProps(tooltip), [tooltip]);

  /** ***************** component *******************/
  return (
    <div
      data-type={style}
      data-size={size}
      data-color={color}
      data-active={isActive}
      data-round={isRound}
      data-block={fullWidth}
      data-square={isSquare}
      data-spinner={!!useSpinner}
      data-loading={isLoading}
      data-disabled={isDisabled}
      data-icon-align={iconAlign}
      data-invert={invertColor}
      className={`ax-button ${className ?? ""}`}
    >
      {hotKey && <AxHotKey keyCombo={hotKey} handler={onClick} />}
      {!useSpinner && isLoading && <div className="ax-button__loader" />}
      <Link
        aria-label={children}
        {...rest}
        type={type}
        as="button"
        role="button"
        ref={ref}
        className="ax-button__inner"
        onClick={handleClick(onClick, { stopPropagation })}
        data-disabled={isDisabled || isLoading}
        data-popover-open={popoverOpen}
        {...tooltipProps}
        tabIndex={noTabFocus ? -1 : 0}
      >
        {!(useSpinner && isLoading) && icon && (
          <AxIcon className="ax-button__icon" rtlFlip={rtlFlip} icon={icon} />
        )}
        {useSpinner && isLoading && (
          <AxAnimation.Spinner className="ax-button__icon" />
        )}
        {children && (
          <Ellipsis className="ax-button__label">{children}</Ellipsis>
        )}
        {hotKey && <AxHotKey.Label keyCombo={hotKey} />}
        {showCaret && (
          <AxIcon className="ax-button__caret" icon={AppIcons.iconCaretDown} />
        )}
      </Link>
      {Badge}
      {extra}
    </div>
  );
}) as AnyObject;
AxButton.Action = ActionButton;
AxButton.Confirm = ConfirmButton;
AxButton.Dropdown = DropdownButton;
AxButton.Group = ({
  children,
  isVertical = false,
  className = "",
}: { isVertical?: boolean } & ElementProps & ChildrenProp) => (
  <div
    className={`ax-button__group ${className ?? ""}`}
    data-vertical={isVertical}
  >
    {children}
  </div>
);

AxButton.displayName = "AxButton";
AxButton.Action.displayName = "AxButton.Action";
AxButton.Confirm.displayName = "AxButton.Confirm";
AxButton.Dropdown.displayName = "AxButton.Dropdown";
AxButton.Group.displayName = "AxButton.Group";
