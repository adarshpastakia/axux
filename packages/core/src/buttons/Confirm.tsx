/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */

import { Popover } from "@headlessui/react";
import { Placement } from "@popperjs/core";
import { FC } from "react";
import { useTranslation } from "react-i18next";
import { AxPopover } from "../overlays/Popover";
import { BooleanCallback } from "../types";
import { AxButton, ButtonProps } from "./Button";

export interface ConfirmProps extends Omit<ButtonProps, "onClick"> {
  /**
   * action message
   */
  message: string;
  /**
   * action type
   */
  actionType?: "danger" | "success";
  /**
   * ok label
   */
  okLabel?: string;
  /**
   * ok label
   */
  cancelLabel?: string;
  /**
   * popover placement
   */
  placement?: Placement;
  /**
   * click handler
   */
  onClick?: BooleanCallback;
}

export const ConfirmButton: FC<ConfirmProps> = ({
  onClick,
  message,
  actionType = "success",
  okLabel,
  cancelLabel,
  placement,
  children,
  ...props
}) => {
  const { t } = useTranslation("core");
  /******************* component *******************/
  return (
    <AxPopover placement={placement}>
      <AxButton {...props}>{children}</AxButton>
      <div className="p-2">
        <p>{message}</p>
        <div>
          <Popover.Button>
            <AxButton
              size="sm"
              style="link"
              color={actionType}
              onClick={() => onClick?.(false)}
            >
              {cancelLabel ?? t("action.cancel")}
            </AxButton>
          </Popover.Button>
          <Popover.Button>
            <AxButton
              size="sm"
              style="solid"
              color={actionType}
              onClick={() => onClick?.(true)}
            >
              {okLabel ?? t("action.ok")}
            </AxButton>
          </Popover.Button>
        </div>
      </div>
    </AxPopover>
  );
};
