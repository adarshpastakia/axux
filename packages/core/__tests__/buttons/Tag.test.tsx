// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2022
// @license   : MIT

import { composeStories } from "@storybook/testing-react";
import { act, fireEvent, render, waitFor } from "@testing-library/react";
import * as stories from "../../__stories__/buttons/TagStory";

const { TagStory, WithIconStory, WithTooltipStory } = composeStories(stories);

describe("Tag", () => {
  it("renders basic tag", (done) => {
    const fragment = render(<TagStory />);
    expect(fragment.container).toHaveTextContent("Tag");
    expect(fragment.container.querySelector(".ax-icon")).toBeNull();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders with icon", (done) => {
    const fragment = render(<WithIconStory />);
    expect(fragment.container).toHaveTextContent("Icon");
    expect(fragment.container.querySelector(".ax-tag__inner > .ax-icon")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders color", (done) => {
    const fragment = render(<TagStory color="secondary" />);
    expect(fragment.container).toHaveTextContent("Tag");
    expect(fragment.container.querySelector(".ax-color--secondary")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    done();
  });

  it("renders style", (done) => {
    const fragment = render(<TagStory fillColor />);
    expect(fragment.container).toHaveTextContent("Tag");
    expect(fragment.container.querySelector("[data-solid='true']")).toBeInTheDocument();
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
    done();
  });

  it("renders with tooltip", async () => {
    const fragment = render(<WithTooltipStory />);
    expect(fragment.container).toHaveTextContent("Tooltip");

    const button = fragment.container.querySelector(".ax-tag__inner") as HTMLElement;
    act(() => {
      fireEvent.mouseOver(button);
    });
    await waitFor(() => fragment.container.querySelector(".ax-tooltip"));
    expect(fragment.container).toHaveTextContent("Test tooltip");
    expect(fragment.container).toMatchSnapshot();
    act(() => {
      fireEvent.mouseOut(button);
    });
    expect(fragment.container).not.toHaveTextContent("Test tooltip");
    expect(fragment.container).toMatchSnapshot();
    fragment.unmount();
  });

  it("should trigger click event", (done) => {
    const handler = jest.fn();
    const fragment = render(<TagStory onClick={handler}>Click</TagStory>);
    fireEvent.click(fragment.getByText("Click"));
    expect(handler).toBeCalled();
    fragment.unmount();
    done();
  });

  it("should trigger remove event", (done) => {
    const handler = jest.fn();
    const fragment = render(<TagStory onRemove={handler}>Click</TagStory>);
    fireEvent.click(fragment.getByRole("remove"));
    expect(handler).toBeCalled();
    fragment.unmount();
    done();
  });
});
