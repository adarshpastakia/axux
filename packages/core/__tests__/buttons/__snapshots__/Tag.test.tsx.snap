// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag renders basic tag 1`] = `
<div>
  <div
    class="ax-tag  ax-color--default"
    data-clickable="false"
    data-size="normal"
    data-solid="false"
  >
    <div
      class="ax-tag__inner"
    >
      <span>
        <span
          class="ax-ellipsis--middle"
          data-enabled="false"
        >
          <span>
            Tag
          </span>
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`Tag renders color 1`] = `
<div>
  <div
    class="ax-tag  ax-color--secondary"
    data-clickable="false"
    data-size="normal"
    data-solid="false"
  >
    <div
      class="ax-tag__inner"
    >
      <span>
        <span
          class="ax-ellipsis--middle"
          data-enabled="false"
        >
          <span>
            Tag
          </span>
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`Tag renders style 1`] = `
<div>
  <div
    class="ax-tag  ax-color--default"
    data-clickable="false"
    data-size="normal"
    data-solid="true"
  >
    <div
      class="ax-tag__inner"
    >
      <span>
        <span
          class="ax-ellipsis--middle"
          data-enabled="false"
        >
          <span>
            Tag
          </span>
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`Tag renders with icon 1`] = `
<div>
  <div
    class="ax-tag  ax-color--default"
    data-clickable="false"
    data-size="normal"
    data-solid="false"
  >
    <div
      class="ax-tag__inner"
    >
      <div
        class="ax-icon "
      >
        <i
          class="mdi mdi-bell"
        />
      </div>
      <span>
        <span
          class="ax-ellipsis--middle"
          data-enabled="false"
        >
          <span>
            Icon
          </span>
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`Tag renders with tooltip 1`] = `
<div>
  <div
    class="ax-tag  ax-color--default"
    data-clickable="false"
    data-size="normal"
    data-solid="false"
  >
    <div
      class="ax-tag__inner"
    >
      <span>
        <span
          class="ax-ellipsis--middle"
          data-enabled="false"
        >
          <span>
            Tooltip
          </span>
        </span>
      </span>
    </div>
  </div>
  <div
    class="ax-popper ax-root ax-tooltip ax-bg--darkest ax-color--empty "
    data-show="true"
    data-show-arrow="true"
    style="position: fixed; left: 0px; top: 0px; min-width: 0;"
  >
    <div
      class="ax-popper--body"
    >
      <span>
        Test tooltip
      </span>
    </div>
    <div
      class="ax-popper--arrow"
      style="position: absolute;"
    />
  </div>
</div>
`;

exports[`Tag renders with tooltip 2`] = `
<div>
  <div
    class="ax-tag  ax-color--default"
    data-clickable="false"
    data-size="normal"
    data-solid="false"
  >
    <div
      class="ax-tag__inner"
    >
      <span>
        <span
          class="ax-ellipsis--middle"
          data-enabled="false"
        >
          <span>
            Tooltip
          </span>
        </span>
      </span>
    </div>
  </div>
</div>
`;
